### PR TITLE
Enhancements to SAML service to support signed logout requests

### PR DESCRIPTION
**Pull request by mayoral**
This PR implements #12355.

It adds some new parameters to `cartodb:saml:create_saml_configuration` in order to help to configure an organization with support for signed logout requests.

Typical unsigned logout request rake task example:

```
ORGANIZATION_NAME=myorg \
SAML_IDP_METADATA_FILE=https://adfs.cartodb.com/federationmetadata/2007-06/federationmetadata.xml \
SAML_EMAIL_ATTRIBUTE=http://schemas.xmlsoap.org/ws/2005/05/identity/claims/emailaddress \
bundle exec rake cartodb:saml:create_saml_configuration
```

Same organization, but activating support for signed logout request, using the default SHA256 digest method and RSA-SHA256 signature method (ex. ADFS):

```
ORGANIZATION_NAME=myorg \
SAML_IDP_METADATA_FILE=https://adfs.cartodb.com/federationmetadata/2007-06/federationmetadata.xml \
SAML_EMAIL_ATTRIBUTE=http://schemas.xmlsoap.org/ws/2005/05/identity/claims/emailaddress \
SAML_SP_PRIVATE_KEY_FILE=sp_private.key \
SAML_SP_CERTIFICATE_FILE=sp_certificate..pem \
SAML_NAME_IDENTIFIER_FORMAT=urn:oasis:names:tc:SAML:1.1:nameid-format:emailAddress \
bundle exec rake cartodb:saml:create_saml_configuration
```

### CODE REVIEW

File: lib/carto/saml_service.rb [hunk 1]
```diff
@@ -109,12 +110,19 @@ def email_attribute
...
       settings = OneLogin::RubySaml::Settings.new
+
       # Make validations throw an error
       settings.soft = false
+
       settings_hash.each do |k, v|
-        method = "#{k}="
-        settings.__send__(method, v) if settings.respond_to?(method)
+        if k.to_s == 'security'
+          settings.security = v
```
Comment (javitonino) [inline lib/carto/saml_service.rb]: 🤔 This looks like it does exactly what the `else` branch of the code. Also, if you do not specify the whole `security` dictionary in the DB, you will remove the default keys from it. I am not sure if it could break anything, but I think something like a `merge` will be better.

File: lib/tasks/saml.rake [hunk 2]
```diff
@@ -36,15 +43,29 @@ namespace :cartodb do
                           name_identifier_format: ENV['SAML_NAME_IDENTIFIER_FORMAT']
                         }
                         config[:idp_slo_target_url] = ENV['SAML_IDP_SLO_TARGET_URL'] if ENV['SAML_IDP_SLO_TARGET_URL'].present?
+
                         config
                       end
 
+      if ENV['SAML_SIGNED_LOGOUT'].present? && ENV['SAML_SIGNED_LOGOUT'] == 'true'
```
Comment (javitonino) [inline lib/tasks/saml.rake]: This may need to check for presence of the following keys:
- `name_identifier_format`
- `idp_slo_target_url`

Those two are required for logout to work as well.

File: lib/tasks/saml.rake [hunk 3]
```diff
@@ -36,15 +43,29 @@ namespace :cartodb do
...
+        configuration[:security] = {
+          logout_requests_signed: true,
+          logout_responses_signed: true,
+          digest_method: ENV['SAML_SLO_DIGEST_METHOD'] || XMLSecurity::Document::SHA256,
+          signature_method: ENV['SAML_SLO_SIGNATURE_METHOD'] || XMLSecurity::Document::RSA_SHA256,
+          embed_sign: true
```
Comment (javitonino) [inline lib/tasks/saml.rake]: I think this is not needed, it's just something that I wrote during testing. Can we try to remove it?

File: lib/tasks/saml.rake [hunk 4]
```diff
@@ -36,15 +43,29 @@ namespace :cartodb do
...
+        configuration[:certificate] = File.read(ENV['SAML_SP_CERTIFICATE']) if ENV['SAML_SP_CERTIFICATE'].present?
+      end
+
       base_url = CartoDB.base_url(organization.name)
       configuration[:issuer] = ENV['SAML_ISSUER'] || base_url + '/saml/metadata'
       configuration[:email_attribute] = ENV['SAML_EMAIL_ATTRIBUTE']
       configuration[:assertion_consumer_service_url] = ENV['SAML_ASSERTION_CONSUMER_SERVICE_URL'] || base_url + '/saml/finalize'
       configuration[:single_logout_service_url] = ENV['SAML_SINGLE_LOGOUT_SERVICE_URL'] || base_url + '/logout'
-      configuration[:name_identifier_format] ||= 'urn:oasis:names:tc:SAML:2.0:nameid-format:transient'
+      configuration[:name_identifier_format] =  ENV['SAML_NAME_IDENTIFIER_FORMAT'] || 'urn:oasis:names:tc:SAML:2.0:nameid-format:transient'
```
Comment (javitonino) [inline lib/tasks/saml.rake]: Why this change? This assignment is done in https://github.com/CartoDB/cartodb/pull/12544/files#diff-d495f91e7229da20f1c405866455d1eeR43, isn't it?

Also, we should probably default to `urn:oasis:names:tc:SAML:1.1:nameid-format:emailAddress`, as we are sending email addresses in the logout request.

File: lib/tasks/saml.rake [hunk 5]
```diff
@@ -36,15 +43,29 @@ namespace :cartodb do
...
+      if ENV['SAML_SIGNED_LOGOUT'].present? && ENV['SAML_SIGNED_LOGOUT'] == 'true'
+        configuration[:security] = {
+          logout_requests_signed: true,
+          logout_responses_signed: true,
+          digest_method: ENV['SAML_SLO_DIGEST_METHOD'] || XMLSecurity::Document::SHA256,
+          signature_method: ENV['SAML_SLO_SIGNATURE_METHOD'] || XMLSecurity::Document::RSA_SHA256,
+          embed_sign: true
+        }
+
+        configuration[:private_key] = File.read(ENV['SAML_SP_PRIVATE_KEY']) if ENV['SAML_SP_PRIVATE_KEY'].present?
```
Comment (javitonino) [inline lib/tasks/saml.rake]: Minor: I'd make it more obvious that this is a keyfile: `SAML_SP_PRIVATE_KEY_FILE`?

File: lib/tasks/saml.rake [hunk 6]
```diff
@@ -7,6 +7,12 @@ namespace :cartodb do
     # SAML_EMAIL_ATTRIBUTE: attribute with the user email. Example: 'email'.
     # SAML_ASSERTION_CONSUMER_SERVICE_URL: [OPTIONAL] CARTO URL for SAML, including organization name. Examples: 'http://192.168.20.2/user/orgname/saml/finalize'. Defaults to the URL built from configuration and organization name
     # SAML_SINGLE_LOGOUT_SERVICE_URL: [OPTIONAL] CARTO URL for SAML logout, including organization name. Examples: 'http://192.168.20.2/user/orgname/logout'. Defaults to the URL built from configuration and organization name
+    # SAML_SIGNED_LOGOUT: [OPTIONAL] If set to 'true', activates support for SAML signed logout request
+    # SAML_SLO_DIGEST_METHOD: [OPTIONAL] Digest method to use in signed logout request. By default: 'http://www.w3.org/2001/04/xmlenc#sha256'
+    # SAML_SLO_SIGNATURE_METHOD: [OPTIONAL] Signature method to use in singled logout requests. By default: 'http://www.w3.org/2001/04/xmldsig-more#rsa-sha256'
+    # SAML_SP_PRIVATE_KEY: [OPTIONAL] Private key file used for signed logout requests
+    # SAML_SP_CERTIFICATE: [OPTIONAL] Certificate file used for signed logout requests
```
Comment (javitonino) [inline lib/tasks/saml.rake]: Please add that it must be `.pem` formatted.

File: lib/tasks/saml.rake [hunk 7]
```diff
@@ -7,6 +7,12 @@ namespace :cartodb do
     # SAML_EMAIL_ATTRIBUTE: attribute with the user email. Example: 'email'.
     # SAML_ASSERTION_CONSUMER_SERVICE_URL: [OPTIONAL] CARTO URL for SAML, including organization name. Examples: 'http://192.168.20.2/user/orgname/saml/finalize'. Defaults to the URL built from configuration and organization name
     # SAML_SINGLE_LOGOUT_SERVICE_URL: [OPTIONAL] CARTO URL for SAML logout, including organization name. Examples: 'http://192.168.20.2/user/orgname/logout'. Defaults to the URL built from configuration and organization name
+    # SAML_SIGNED_LOGOUT: [OPTIONAL] If set to 'true', activates support for SAML signed logout request
```
Comment (javitonino) [inline lib/tasks/saml.rake]: I would remove this parameter, and just activate/deactivate signed logout based on the presence of the key & certificate parameters.

**Comment by javitonino** [comment]
Did you get a chance to look at this part of the ticket?
> As a bonus, since this increase configuration difficulty: we should test that the metadata file works for autoconfiguring with multiple providers. Right now, it works with simplesamlphp, but it doesn't work very well with ADFS, since it insists in adding federationmetadata.xml to the URL for some reason (probably because our route doesn't end in .xml).

Maybe worth a look, since IdP config is becoming more complicated. It's minor, anyway. If we don't do it, I would like to a see a PR against our documentation for ADFS configuration.

File: lib/carto/saml_service.rb (hunk 1 above)
Comment (mayoral) [inline lib/carto/saml_service.rb]: Changed to a merge so we don't overwrite the whole security hash.

File: lib/tasks/saml.rake (hunk 7 above)
Comment (mayoral) [inline lib/tasks/saml.rake]: I've removed the param, and changed to check four needed params.

File: lib/tasks/saml.rake (hunk 6 above)
Comment (mayoral) [inline lib/tasks/saml.rake]: Done.

File: lib/tasks/saml.rake (hunk 2 above)
Comment (mayoral) [inline lib/tasks/saml.rake]: Done.

File: lib/tasks/saml.rake (hunk 3 above)
Comment (mayoral) [inline lib/tasks/saml.rake]: I think I did my tests yesterday with embed true and false, and there was no difference :/

File: lib/tasks/saml.rake (hunk 5 above)
Comment (mayoral) [inline lib/tasks/saml.rake]: Done, and I've also added the _FILE to the certificate env var.

File: lib/tasks/saml.rake (hunk 4 above)
Comment (mayoral) [inline lib/tasks/saml.rake]: You're right, the assignment was done earlier. Changed to default to emailAddress.

**Comment by mayoral** [comment]
@javitonino I've just seen your comment about the bonus part. I didn't take a look to that part of the issue, my focus was to get the signed logout working :(

File: lib/carto/saml_service.rb [hunk 8]
```diff
@@ -109,12 +110,19 @@ def email_attribute
...
       settings = OneLogin::RubySaml::Settings.new
+
       # Make validations throw an error
       settings.soft = false
+
       settings_hash.each do |k, v|
-        method = "#{k}="
-        settings.__send__(method, v) if settings.respond_to?(method)
+        if k.to_s == 'security'
+          settings.security.merge(v)
```
Comment (javitonino) [inline lib/carto/saml_service.rb]: Does this work? IIRC `merge` doesn't modify the hash. Maybe `merge!`?

File: lib/tasks/saml.rake [hunk 9]
```diff
@@ -36,15 +42,29 @@ namespace :cartodb do
...
+        configuration[:certificate] = File.read(ENV['SAML_SP_CERTIFICATE_FILE'])
+      end
+
       base_url = CartoDB.base_url(organization.name)
       configuration[:issuer] = ENV['SAML_ISSUER'] || base_url + '/saml/metadata'
       configuration[:email_attribute] = ENV['SAML_EMAIL_ATTRIBUTE']
       configuration[:assertion_consumer_service_url] = ENV['SAML_ASSERTION_CONSUMER_SERVICE_URL'] || base_url + '/saml/finalize'
       configuration[:single_logout_service_url] = ENV['SAML_SINGLE_LOGOUT_SERVICE_URL'] || base_url + '/logout'
-      configuration[:name_identifier_format] ||= 'urn:oasis:names:tc:SAML:2.0:nameid-format:transient'
+      configuration[:name_identifier_format] ||= 'urn:oasis:names:tc:SAML:2.0:nameid-format:emailAddress'
```
Comment (javitonino) [inline lib/tasks/saml.rake]: Nope. It is `urn:oasis:names:tc:SAML:1.1:nameid-format:emailAddress` (SAML:1.1)

File: lib/tasks/saml.rake [hunk 10]
```diff
@@ -36,15 +42,29 @@ namespace :cartodb do
                           name_identifier_format: ENV['SAML_NAME_IDENTIFIER_FORMAT']
                         }
                         config[:idp_slo_target_url] = ENV['SAML_IDP_SLO_TARGET_URL'] if ENV['SAML_IDP_SLO_TARGET_URL'].present?
+
                         config
                       end
 
+      if ENV['SAML_SP_PRIVATE_KEY_FILE'].present? && ENV['SAML_SP_CERTIFICATE_FILE'].present? &&
+        ENV['SAML_NAME_IDENTIFIER_FORMAT'].present && ENV['SAML_IDP_SLO_TARGET_URL'].present?
```
Comment (javitonino) [inline lib/tasks/saml.rake]: Uhm, have you tested if this change works? I think it adds a lot of unnecessary parameters for autoconfiguration.

Better to check for the actual value in the `config` hash. This is because if you configure with autoconfiguration (federationmetadata.xml), these last two env variables will be null, but the config values will be populated.

File: lib/carto/saml_service.rb (hunk 8 above)
Comment (mayoral) [inline lib/carto/saml_service.rb]: Yep, changed to `merge!`.

File: lib/tasks/saml.rake (hunk 10 above)
Comment (mayoral) [inline lib/tasks/saml.rake]: I've replaced the last two env vars check with a check for `configuration` hash values, so we support both scenarios.

File: lib/tasks/saml.rake (hunk 9 above)
Comment (mayoral) [inline lib/tasks/saml.rake]: Done.

**Comment by javitonino** [comment]
LGTM. @juanignaciosl please take a look (as I wrote some code in this as well).

**Comment by juanignaciosl** [comment]
LGTM. 🚀  to acceptance!

**Comment by javitonino** [comment]
Acceptance failed 🔴 
- I had to correct a typo for the rake to run (https://github.com/CartoDB/cartodb/pull/12544/commits/6615fd2e95f68fc148c4f6bd18e3f13e3aca37f1)
- Although it is configured correctly, CARTO is not sending a signed request to ADFS, it seems the configuration is not being read correctly. Looks like a string/symbols as keys error:

```
(byebug) settings_hash['security']
{"logout_requests_signed"=>true, "logout_responses_signed"=>true, "digest_method"=>"http://www.w3.org/2001/04/xmlenc#sha256", "signature_method"=>"http://www.w3.org/2001/04/xmldsig-more#rsa-sha256"}
(byebug) settings.security
{:authn_requests_signed=>false, :logout_requests_signed=>false, :logout_responses_signed=>false, :want_assertions_signed=>false, :want_name_id=>false, :metadata_signed=>false, :embed_sign=>false, :digest_method=>"http://www.w3.org/2000/09/xmldsig#sha1", :signature_method=>"http://www.w3.org/2000/09/xmldsig#rsa-sha1", "logout_requests_signed"=>true, "logout_responses_signed"=>true, "digest_method"=>"http://www.w3.org/2001/04/xmlenc#sha256", "signature_method"=>"http://www.w3.org/2001/04/xmldsig-more#rsa-sha256"}
```

**Comment by mayoral** [comment]
I've added a `symbolize_keys` for the security hash in saml_settings. Can you try again?

**Comment by javitonino** [comment]
This now works pretty well 🙌 

Only have a complaint, that it is a bit of an edge case, but might be important.
**STR**
1. Login with SAML
2. Remove the cookie from the SAML IdP (adfs.cartodb.com in this example)
3. Logout from CARTO

You will get a 500 in CARTO: 
```
OneLogin::RubySaml::ValidationError (Bad status code. Expected <urn:oasis:names:tc:SAML:2.0:status:Success>, but was: <urn:oasis:names:tc:SAML:2.0:status:Requester>)
```

This is because ADFS returns an error because the user is already logged out. I feel that in this case, we should ignore the error and logout from CARTO nonetheless, maybe logging the SAML error.

**Comment by javitonino** [comment]
Tested and looks good. Logout works in all three tested cases:
- Logout from CARTO, logged in to ADFS
- Logout from CARTO, withotu active session in ADFS
- Logout from ADFS

Scheduling for Monday deploy.

💃 🎉
